xrtl/gfx/util/memory_command_buffer_reader.h

```diff
@@ -0,0 +1,93 @@
+// Copyright 2017 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef XRTL_GFX_UTIL_MEMORY_COMMAND_BUFFER_READER_H_
+#define XRTL_GFX_UTIL_MEMORY_COMMAND_BUFFER_READER_H_
+
+#include "xrtl/base/array_view.h"
+#include "xrtl/base/ref_ptr.h"
+#include "xrtl/gfx/util/memory_commands.h"
+
+namespace xrtl {
+namespace gfx {
+namespace util {
+
+// Reads data from a command buffer in the form written by
+// MemoryCommandBufferWriter.
+class MemoryCommandBufferReader {
+ public:
+  explicit MemoryCommandBufferReader(const PacketHeader* packet)
+      : current_packet_(packet) {}
+
+  // Returns true if the reader is empty and there is no more data remaining in
+  // the command buffer.
+  bool empty() const;
+
+  // Peeks at a command header in the command buffer.
+  // Returns nullptr if there is no valid command or the buffer is empty.
+  const CommandHeader* PeekCommandHeader();
+
+  // Reads a command struct from the command buffer.
+  template <typename T>
+  const T& ReadCommand(const CommandHeader* command_header) {
+    packet_offset_ += sizeof(CommandHeader) + sizeof(T);
+    return *reinterpret_cast<const T*>(
+        reinterpret_cast<const uint8_t*>(command_header) +
+        sizeof(CommandHeader));
+  }
+
+  // Reads a raw data blob from the command buffer.
+  const void* ReadData(size_t data_length);
+
+  // Reads an array of primitives/structs from the command buffer.
+  template <typename T>
+  ArrayView<T> ReadArray(size_t value_count) {
+    if (!value_count) {
+      return {};
+    }
+    return ArrayView<T>(
+        reinterpret_cast<const T*>(ReadData(value_count * sizeof(T))),
```

Comment:
that's an optimization only to avoid making the full ReadData call for zero size arrays - it's not required for correctness